app/master/cluster_master.py

```diff
@@ -228,7 +224,22 @@ def handle_request_for_new_build(self, build_params):
         if build_request.is_valid():
             build = Build(build_request)
             self._all_builds_by_id[build.build_id()] = build
-            self._request_queue.put(build)
+
+            build.generate_project_type()
+            project_id = build.project_type.project_id()
+            if project_id in self._request_queues.keys():
+                self._request_queues[project_id].put(build)
+            else:
+                self._request_queues[project_id] = Queue()
+                self._request_queues[project_id].put(build)
```

Comment:
I know this is just a WIP, but I believe handle_request_for_new_build can be called concurrently, meaning that there is a chance that you will have multiple build requests for the same project coming in, and consequently we would overwrite the same project_id in the self._request_queue queue.